### DISCUSSION TITLE
trying to update a many to many relation sends invalid value error

### DISCUSSION THREAD

**Discussion by Mazeral** [prisma/prisma#14284] [Q&A] [answered]
I have a many to many relation between two models, this my schema to show the relation:
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Author {
  id  Int @id @default(autoincrement())
  FirstName String 
  LastName String
  books Book[]
}

model Book{
  id Int @id @default(autoincrement())
  Title String
  Description String
  authors Author[]
}
```
I'm trying to make an update function that can update the `books` in my author like this:
```ts
 return this.prisma.author.update({
        where: {
          id: id,
        },
        data: {
          books: cleanarray,
        },
      });
```
But when I try to send the request, it throws this error:
```
      Error: 
Invalid `this.prisma.author.update()` invocation in
/home/mohammadomar/training/library-dojo/src/author/author.service.ts:111:33

  108 // Has been fixed.
  109 const id: number = author.id;
  110 console.log(id + 'this is the id');
→ 111 return this.prisma.author.update({
        where: {
          id: 2
        },
        data: {
          books: [
            {
              Title: 'title1',
              Description: 'desc1'
            },
            {
              Title: 'title2',
              Description: 'desc2'
            },
            {
              Title: 'title3',
              Description: 'desc3'
            }
          ]
          ~~~~~~~~~~~~~~~~~~~~~~~~
        }
      })

Argument books: Got invalid value 
[
  {
    Title: 'title1',
    Description: 'desc1'
  },
  {
    Title: 'title2',
    Description: 'desc2'
  },
  {
    Title: 'title3',
    Description: 'desc3'
  }
]
on prisma.updateOneAuthor. Provided List<Json>, expected BookUpdateManyWithoutAuthorsNestedInput:
type BookUpdateManyWithoutAuthorsNestedInput {
  create?: BookCreateWithoutAuthorsInput | List<BookCreateWithoutAuthorsInput> | BookUncheckedCreateWithoutAuthorsInput | List<BookUncheckedCreateWithoutAuthorsInput>
  connectOrCreate?: BookCreateOrConnectWithoutAuthorsInput | List<BookCreateOrConnectWithoutAuthorsInput>
  upsert?: BookUpsertWithWhereUniqueWithoutAuthorsInput | List<BookUpsertWithWhereUniqueWithoutAuthorsInput>
  set?: BookWhereUniqueInput | List<BookWhereUniqueInput>
  disconnect?: BookWhereUniqueInput | List<BookWhereUniqueInput>
  delete?: BookWhereUniqueInput | List<BookWhereUniqueInput>
  connect?: BookWhereUniqueInput | List<BookWhereUniqueInput>
  update?: BookUpdateWithWhereUniqueWithoutAuthorsInput | List<BookUpdateWithWhereUniqueWithoutAuthorsInput>
  updateMany?: BookUpdateManyWithWhereWithoutAuthorsInput | List<BookUpdateManyWithWhereWithoutAuthorsInput>
  deleteMany?: BookScalarWhereInput | List<BookScalarWhereInput>
}

```

I've checked the documentations that talks about querying with many to many relations but it didn't really help me out

**Comment by Mazeral** [comment]
I tried to make the book object contain the `title`, `description` and `authors[]`, it gave me the same results.
I tried to add to them the `id` too, didn't work.

**Comment by Mazeral** [comment]
I tried to use update many but this error appeard:
```Type '{ books: any; }' is not assignable to type '(Without<BookUpdateManyMutationInput, BookUncheckedUpdateManyWithoutBooksInput> & BookUncheckedUpdateManyWithoutBooksInput) | (Without<...> & BookUpdateManyMutationInput)'.
  Object literal may only specify known properties, and 'books' does not exist in type '(Without<BookUpdateManyMutationInput, BookUncheckedUpdateManyWithoutBooksInput> & BookUncheckedUpdateManyWithoutBooksInput) | (Without<...> & BookUpdateManyMutationInput)'.

```

**Comment by ruheni** [comment, marked as answer]
Hey 👋🏽

It seems your query is missing the `update`/ `updateMany` property in the `books property. Here are 2 queries you could try out to resolve your error: 

- Updating a single object:
```ts
await prisma.author.update({
  where: {
    id: id
  },
  data: {
    books: {
      update: {
        data: { /** title and description here */ },
        where: {/** book's filters here */ }
      }
    }
  }

```

- Updating multiple (an array) objects: 
```ts
await prisma.author.update({
  where: {
    id: 1
  },
  data: {
    books: {
      updateMany: {
        data: [],
        where: {/** book's filters here */ }
      }
    }
  }
})
```

We have some docs on [working with many-to-many relationships](https://www.prisma.io/docs/guides/database/troubleshooting-orm/help-articles/working-with-many-to-many-relations#implicit-relations) that includes a snippet of how you can create an update query too. 🙂

Let me know if you hit a snag.

**Reply by Mazeral** [reply]
The author currently doesn't have any book elements in its array, what should the `where` object contain?
I tried to just not write it because I didn't really have any book data to filter but it gave me this error 
```
Type '{ data: Book[]; }' is not assignable to type 'Enumerable<BookUpdateManyWithWhereWithoutAuthorsInput>'.
  Property 'where' is missing in type '{ data: Book[]; }' but required in type 'BookUpdateManyWithWhereWithoutAuthorsInput'.ts(2322)
index.d.ts(3343, 5): 'where' is declared here.
index.d.ts(3204, 5): The expected type comes from property 'updateMany' which is declared here on type 'BookUncheckedUpdateManyWithoutAuthorsNestedInput | BookUpdateManyWithoutAuthorsNestedInput'
```

**Reply by ruheni** [reply]
> The author currently doesn't have any book elements in its array, what should the `where` object contain?

If I understand correctly, would you like to (a) create book record(s) by an author's `id`? If so, you can use either `create`  or `createMany`.

**Reply by Mazeral** [reply]
I though that the author has an empty array of books which is declared in the model, am I wrong?

**Reply by ruheni** [reply]
When you create a new author record, no book records will be attached to it — unless you define it in the nested write when creating an author record.i.e

```ts
// no books will be attached to the author record here
await prisma.author.create({
  data: {
    title: '',
     description: ''
  }
})
```

**Reply by Mazeral** [reply]
I'm so sorry but I just, couldn't do it...I really tried but I couldn't make it work
this is the author structure which I want to add the array of books:
I used findMany() without any `select :true`
```json
[
    {
        "id": 2,
        "FirstName": "author 1",
        "LastName": "last 1"
    }
]
```

and these are the books that i want to insert them in the `Book[]` array:
```ts
[
  { id: 6, Title: 'title1', Description: 'desc1' },
  { id: 7, Title: 'title2', Description: 'desc2' },
  { id: 8, Title: 'title3', Description: 'desc3' }
]

```
Can you help me figuring this out?

**Reply by pantharshit00** [reply]
Hello @Mazeral, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

You can do the follow query here that will update the `Author` record and it will create and associate book with it:

```ts
  await prisma.author.update({
    where: {
      id: 1,
    },
    data: {
      books: {
        create: cleanArray,
      },
    },
  });

```

**Reply by Mazeral** [reply]
Thanks for caring, I've solved this solution by using a for each loop

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
